tedana/utils/utils.py

```diff
@@ -1,198 +1,452 @@
+    fdata : (S x T) np.ndarray
+        Reshaped `data`, where `S` is samples and `T` is time
     """
-    nx, ny, nz, Ne, _ = cdat.shape
 
-    mask = np.ones((nx, ny, nz), dtype=np.bool)
+    if isinstance(data, str):
+        if get_dtype(data) == 'GIFTI':
+            fdata = np.column_stack([f.data for f in nib.load(data).darrays])
+            return fdata
+        elif get_dtype(data) == 'NIFTI':
+            data = check_niimg(data).get_data()
 
-    if minimum:
-        mask = cdat.prod(axis=-1).prod(-1) != 0
-        return mask
-    else:
-        # Make a map of longest echo that a voxel can be sampled with,
-        # with minimum value of map as X value of voxel that has median
-        # value in the 1st echo. N.b. larger factor leads to bias to lower TEs
-        emeans = cdat.mean(-1)
-        medv = emeans[:, :, :, 0] == scoreatpercentile(emeans[:, :, :, 0][emeans[:, :, :, 0] != 0],
-                                                       33, interpolation_method='higher')
-        lthrs = np.squeeze(np.array([emeans[:, :, :, ee][medv] / 3 for ee in range(Ne)]))
-
-        if len(lthrs.shape) == 1:
-            lthrs = np.atleast_2d(lthrs).T
-        lthrs = lthrs[:, lthrs.sum(0).argmax()]
-
-        mthr = np.ones([nx, ny, nz, Ne])
-        for i_echo in range(Ne):
-            mthr[:, :, :, i_echo] *= lthrs[i_echo]
-        mthr = np.abs(emeans) > mthr
-        masksum = np.array(mthr, dtype=np.int).sum(-1)
-        mask = masksum != 0
-        if getsum:
-            return mask, masksum
-        else:
-            return mask
+    fdata = data.reshape((-1,) + data.shape[3:])
+
+    return fdata.squeeze()
 
 
-def uncat2echos(data, Ne):
+def load_data(data, n_echos=None):
     """
-    Combines z- and echo-axis in `data`
+    Coerces input `data` files to required 3D array output
 
     Parameters
     ----------
-    data : array_like
-        Array of shape (nx, ny, nz, Ne, nt)
-    Ne : int
-        Number of echoes; should be equal to `data.shape[3]`
+    data : (X x Y x M x T) array_like or list-of-img_like
+        Input multi-echo data array, where `X` and `Y` are spatial dimensions,
+        `M` is the Z-spatial dimensions with all the input echos concatenated,
+        and `T` is time. A list of image-like objects (e.g., .nii or .gii) are
+        accepted, as well
+    n_echos : int, optional
+        Number of echos in provided data array. Only necessary if `data` is
+        array_like. Default: None
 
     Returns
     -------
-    ndarray
-        Array of shape (nx, ny, nz*Ne, nt)
+    fdata : (S x E x T) np.ndarray
+        Output data where `S` is samples, `E` is echos, and `T` is time
+    ref_img : str
+        Filepath to reference image for saving output files
     """
 
-    nx, ny = data.shape[0:2]
-    nz = data.shape[2] * Ne
-    if len(data.shape) > 4:
-        nt = data.shape[4]
-    else:
-        nt = 1
-    return np.reshape(data, (nx, ny, nz, nt), order='F')
+    if isinstance(data, list):
```

Comment:
Should we raise an error for the `else` condition, here? I'm not sure how it could be triggered, but just in case....

Reply:
This is worth some discussion about how we envision people using the tedana interface. As it stands, the primary mechanism for doing that is through `argparse` in `cli/run.py`; if that's the case, `data` will *always* be a list. However, if the goal is to make this a bit more modular, then we should probably add in some other checks. 

Right now even if `data` is a `str` pointing to a nifti file this function should still work (without an explicit `else` clause). But it might be good to enforce some requirements! I'm open to comments/suggestions and will be happy to build them in.

Reply:
Yes, I agree there's a broader discussion to have about how to use the tedana interface-- I'll punt on that for now. It sounds like this can handle the two most likely cases, so we can revisit when it comes time to talk about how we expect people to run the code.

Reply:
It's unclear how necessary this is, but I'm fairly sure some of the current component selection criteria use spatial information. For example, in tedana.py, see the current code starting at line 652: "Step 2: Compute 3-D spatial FFT of Beta maps to detect high-spatial
    frequency artifacts"

The simple solution would be to just skip 3-D space dependent criteria with gii files. I suspect that won't cause much harm, but it would require a some close examination of results processed as volumes vs surfaces. 

My hunch is that the spatial transforms and interpolations across voxels to go from a volume to a surface might cause some undesired issues with echo-dependent estimations, but there might also be some issues with non-trivial transforms & interpolations in volume space as well. I also don't have a good intuitive sense of how tedana classification mistakes would look in surface space, so there might need to be some thought on how to best visualize results to help users identify issues.

Reply:
Good point, and thanks for raising it! Yes, there are going to have to be some differences between the pipelines for volumetric + surface data. Thankfully, having CircleCI running will ensure that none of the changes I'm making in this PR impact the volumetric results in any way, for now. 

A lot of this PR is actually just refactoring to ensure that data can be seamlessly passed between functions without regard for the **type** of data being passed. This will hopefully be useful later in making the code more modular.

As for visualization of surface-based results: I completely agree. I think there will have to be some significant testing to ensure that `tedana` is actually suitable for surface data, given the limited spatial information. Hopefully this PR will be the first step in that process!